build_catalog.sh

```diff
@@ -1,105 +1,186 @@
+        )
+
+        log "Current deployed hash is $deployed_hash"
+
+        new_bundle_versions_file=$(mktemp -p "$temp_dir")
+        delete=false
+        # Sort based on commit number
+        for version in $(sort -t . -k 3 -g "$bundle_versions_file"); do
+            if [[ "$delete" == false ]]; then
+                echo "$version" >> "$new_bundle_versions_file"
+
+                short_hash=$(echo "$version" | cut -d- -f2)
+
+                if [[ "$deployed_hash" == "${short_hash}"* ]]; then
+                    log "found deployed hash in the bundle versions repository"
+                    delete=true
+                fi
+            else
+                log "Adding $version to removed versions"
+                removed_versions="$version $removed_versions"
+            fi
+        done
+        [[ -n "$removed_versions" ]] && log "The following versions will be removed: $removed_versions"
+        cp "$new_bundle_versions_file" "$bundle_versions_file"
     fi
-    log "Found previous bundle version $prev_version"
-  popd
-  rm -rf $tmp_dir
-  docker rm tmp_$$
+    prev_operator_version=$(tail -n 1 "$bundle_versions_file")
+else
+    log "No $bundle_versions_file exist. This is the first time the operator is built"
+    new_bundle_versions_file="$bundle_versions_file"
+    touch "$bundle_versions_file"
 fi
 
-# Build/push the new bundle
-pushd deploy/bundle
-  log "Creating bundle $BUNDLE_IMAGE:$current_commit"
-  if [[ $prev_version != "" ]]; then
-    export REPLACE_VERSION=$prev_version
-  fi
-  export BUNDLE_IMAGE_TAG=$current_commit
-  export OPERATOR_IMAGE_TAG=v$version
-  export VERSION=$version
-  make bundle
-  docker tag $BUNDLE_IMAGE:$current_commit $BUNDLE_IMAGE:latest
-
-  log "Pushing the bundle $BUNDLE_IMAGE:$current_commit to repository"
-  $docker_cmd push $BUNDLE_IMAGE:$current_commit
-  # Do not push the latest tag here.  If there is a problem creating the catalog then
-  # pushing the latest tag here will mean subsequent runs will be extracting a bundle
-  # version that isn't referenced in the catalog.  This will result in all future
-  # catalog creation failing to be created.
-popd
-
-# Download opm build
-curl -L https://github.com/operator-framework/operator-registry/releases/download/v$opm_version/linux-amd64-opm -o ./opm
-chmod u+x ./opm
-
-# Create/push a new catalog via opm
-log "Pulling existing latest catalog $CATALOG_IMAGE"
-$docker_cmd pull $CATALOG_IMAGE:latest && exists=1 || exists=0
-if [ $exists -eq 1 ]; then
-  from_arg="--from-index $CATALOG_IMAGE:latest"
+if [[ "$OPERATOR_VERSION" == "$prev_operator_version" ]]; then
+    log "stopping script as $OPERATOR_VERSION version was already built, so no need to rebuild it"
+    exit 0
 fi
 
-if [[ "$from_arg" == "" ]]; then
-  log "Creating new catalog $CATALOG_IMAGE"
-else
-  log "Updating existing catalog $CATALOG_IMAGE"
+log "Creating bundle image $BUNDLE_IMAGE:$CURRENT_COMMIT"
```

Comment:
This should probably call out to a script that the repo can create/maintain.

Reply:
agree... My idea is to have this working and then refactor it to separate in at least two parts

Reply:
Makes sense.